File: packages/typescript/src/drivers/SeleniumCdpConnection.ts
const AUTO_ATTACH_PARAMS = {
  autoAttach: true,
  waitForDebuggerOnStart: true,
  flatten: true,
};
const TIMEOUT_MS = 5000;

interface CdpMessage {
  id?: number;
  method?: string;
  params?: Record<string, unknown>;
  result?: Record<string, unknown>;
  error?: { message?: string };
  sessionId?: string;
}

interface PendingCommand {
  resolve: (result: Record<string, unknown>) => void;
  reject: (error: Error) => void;
  timeout: ReturnType<typeof setTimeout>;
}

export class SeleniumCdpConnection {
  #socket: WebSocket;
  #waiterScript: string;
  #nextId = 1;
  #pending = new Map<number, PendingCommand>();
  #targetSessions = new Map<string, string>();
  #sessionConfigurations = new Map<string, Promise<void>>();
  #closed = false;

  private constructor(socket: WebSocket, waiterScript: string) {
    this.#socket = socket;
    this.#waiterScript = waiterScript;
    socket.addEventListener("message", (event) => this.#onMessage(event.data));
    socket.addEventListener("close", () => this.#failConnection());
    socket.addEventListener("error", () => this.#failConnection());
  }

  static async connect(
    capabilities: { get(key: string): unknown },
    waiterScript: string,
  ): Promise<SeleniumCdpConnection> {
    const url = await websocketUrl(capabilities);
    const socket = new WebSocket(url);
    try {
      await new Promise<void>((resolve, reject) => {
        const timeout = setTimeout(
          () => reject(new Error("Timed out connecting to Chrome CDP")),
          TIMEOUT_MS,
        );
        socket.addEventListener(
          "open",
          () => {
            clearTimeout(timeout);
            resolve();
          },
          { once: true },
        );
        socket.addEventListener(
          "error",
          () => {
            clearTimeout(timeout);
            reject(new Error("Could not connect to Chrome CDP"));
          },
          { once: true },
        );
      });
    } catch (error) {
      socket.close();
      throw error;
    }
    const connection = new SeleniumCdpConnection(socket, waiterScript);
    try {
      await connection.#send("Target.setAutoAttach", AUTO_ATTACH_PARAMS);
      await connection.#send("Target.setDiscoverTargets", { discover: true });
      const targets = (await connection.#send("Target.getTargets")) as {
        targetInfos?: Array<{ targetId: string; type: string }>;
      };
      for (const target of targets.targetInfos ?? []) {
        if (target.type === "page")
          await connection.#awaitSession(target.targetId);
      }
      return connection;
    } catch (error) {
      connection.close();
      throw error;
    }
  }

  close(): void {
    if (this.#closed) return;
    this.#closed = true;
    this.#failPending(new Error("CDP connection closed"));
    this.#socket.close();
    this.#targetSessions.clear();
    this.#sessionConfigurations.clear();
  }

  #send(
    method: string,
    params: object = {},
    sessionId = "",
    wait = true,
  ): Promise<Record<string, unknown>> {
    if (this.#closed) return Promise.reject(new Error("CDP connection closed"));
    const id = this.#nextId++;
    const message = { id, method, params, ...(sessionId ? { sessionId } : {}) };
    if (!wait) {
      this.#socket.send(JSON.stringify(message));
      return Promise.resolve({});
    }
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.#pending.delete(id);
        reject(new Error(`Timed out sending CDP command ${method}`));
      }, TIMEOUT_MS);
      this.#pending.set(id, { resolve, reject, timeout });
      try {
        this.#socket.send(JSON.stringify(message));
      } catch (error) {
        clearTimeout(timeout);
        this.#pending.delete(id);
        reject(error instanceof Error ? error : new Error(String(error)));
      }
    });
  }

  #onMessage(data: string | ArrayBuffer | Blob): void {
    if (data instanceof Blob) {
      void data.text().then((text) => this.#onMessage(text));
      return;
    }
    const text =
      typeof data === "string" ? data : new TextDecoder().decode(data);
    const message = JSON.parse(text) as CdpMessage;
    if (message.id) {
      const command = this.#pending.get(message.id);
      if (!command) return;
      this.#pending.delete(message.id);
      clearTimeout(command.timeout);
      if (message.error) command.reject(new Error(message.error.message));
      else command.resolve(message.result ?? {});
      return;
    }

    const method = message.method ?? "";
    const params = message.params ?? {};
    if (method === "Target.attachedToTarget") {
      const targetInfo = params.targetInfo as
        | { targetId?: string; type?: string }
        | undefined;
      const sessionId = String(params.sessionId);
      let configuration = Promise.resolve();
      if (targetInfo?.type === "page" || targetInfo?.type === "iframe") {
        if (targetInfo.targetId)
          this.#targetSessions.set(targetInfo.targetId, sessionId);
        configuration = this.#configureSession(sessionId);
      }
      void configuration
        .finally(() =>
          this.#send("Runtime.runIfWaitingForDebugger", {}, sessionId),
        )
        .catch(() => undefined);
    } else if (method === "Target.detachedFromTarget") {
      this.#detachSession(String(params.sessionId ?? ""));
    }
  }

  #configureSession(sessionId: string): Promise<void> {
    const existing = this.#sessionConfigurations.get(sessionId);
    if (existing) return existing;
    const configuration = this.#configureNewSession(sessionId);
    this.#sessionConfigurations.set(sessionId, configuration);
    return configuration;
  }

  async #configureNewSession(sessionId: string): Promise<void> {
    await this.#send("Target.setAutoAttach", AUTO_ATTACH_PARAMS, sessionId);
    await this.#send("Page.enable", {}, sessionId);
    await this.#send(
      "Page.addScriptToEvaluateOnNewDocument",
      { source: this.#waiterScript, runImmediately: true },
      sessionId,
    );
  }

  async #awaitSession(targetId: string): Promise<string> {
    const deadline = Date.now() + TIMEOUT_MS;
    while (Date.now() < deadline) {
      const sessionId = this.#targetSessions.get(targetId);
      const configuration = sessionId
        ? this.#sessionConfigurations.get(sessionId)
        : undefined;
      if (sessionId && configuration) {
        await configuration;
        return sessionId;
      }
      await new Promise((resolve) => setTimeout(resolve, 10));
    }
    return "";
  }

  #detachSession(sessionId: string): void {
    if (!sessionId) return;
    this.#sessionConfigurations.delete(sessionId);
    for (const [targetId, attachedSession] of this.#targetSessions) {
      if (attachedSession === sessionId) this.#targetSessions.delete(targetId);
    }
  }

  #failConnection(): void {
    this.#closed = true;
    this.#failPending(new Error("CDP connection closed"));
  }

  #failPending(error: Error): void {
    for (const command of this.#pending.values()) {
      clearTimeout(command.timeout);
      command.reject(error);
    }
    this.#pending.clear();
  }
}

async function websocketUrl(capabilities: {
  get(key: string): unknown;
}): Promise<string> {
  const cdpUrl = capabilities.get("se:cdp");
  if (typeof cdpUrl === "string") return cdpUrl;

  const options = (capabilities.get("goog:chromeOptions") ??
    capabilities.get("ms:edgeOptions")) as
    | { debuggerAddress?: string }
    | undefined;
  if (!options?.debuggerAddress) {
    throw new Error("Chromium did not expose a CDP debugger address");
  }
  const response = await fetch(
    `http://${options.debuggerAddress}/json/version`,
    {
      signal: AbortSignal.timeout(TIMEOUT_MS),
    },
  );
  if (!response.ok) throw new Error(`CDP discovery failed: ${response.status}`);
  const version = (await response.json()) as { webSocketDebuggerUrl: string };
  return version.webSocketDebuggerUrl;
}

File: packages/java/src/main/java/ai/alumnium/driver/SeleniumCdpConnection.java
package ai.alumnium.driver;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.WebSocket;
import java.nio.ByteBuffer;
import java.time.Duration;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;
import org.openqa.selenium.Capabilities;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

final class SeleniumCdpConnection implements AutoCloseable, WebSocket.Listener {
  private static final Logger LOG = LoggerFactory.getLogger(SeleniumCdpConnection.class);
  private static final ObjectMapper MAPPER = new ObjectMapper();
  private static final Duration TIMEOUT = Duration.ofSeconds(5);
  private static final Map<String, Object> AUTO_ATTACH_PARAMS =
      Map.of("autoAttach", true, "waitForDebuggerOnStart", true, "flatten", true);

  private final String waiterScript;
  private final AtomicLong nextId = new AtomicLong(1);
  private final ConcurrentMap<Long, CompletableFuture<JsonNode>> pending =
      new ConcurrentHashMap<>();
  private final ConcurrentMap<String, String> targetSessions = new ConcurrentHashMap<>();
  private final ConcurrentMap<String, CompletableFuture<Void>> sessionConfigurations =
      new ConcurrentHashMap<>();
  private final AtomicBoolean closed = new AtomicBoolean();
  private final Object textLock = new Object();
  private final Object sendLock = new Object();
  private final StringBuilder textMessage = new StringBuilder();

  private volatile WebSocket socket;
  private CompletableFuture<Void> sendTail = CompletableFuture.completedFuture(null);

  private SeleniumCdpConnection(String waiterScript) {
    this.waiterScript = waiterScript;
  }

  static SeleniumCdpConnection connect(Capabilities capabilities, String waiterScript) {
    SeleniumCdpConnection connection = new SeleniumCdpConnection(waiterScript);
    try {
      URI url = websocketUrl(capabilities);
      HttpClient client = HttpClient.newBuilder().connectTimeout(TIMEOUT).build();
      connection.socket =
          client.newWebSocketBuilder().connectTimeout(TIMEOUT).buildAsync(url, connection).join();

      connection.send("Target.setAutoAttach", AUTO_ATTACH_PARAMS).join();
      connection.send("Target.setDiscoverTargets", Map.of("discover", true)).join();
      JsonNode targets = connection.send("Target.getTargets", Map.of()).join().path("targetInfos");
      if (targets.isArray()) {
        for (JsonNode target : targets) {
          if ("page".equals(target.path("type").asText())) {
            connection.awaitSession(target.path("targetId").asText());
          }
        }
      }
      return connection;
    } catch (RuntimeException e) {
      connection.close();
      throw connectionFailure(e);
    }
  }

  private String awaitSession(String targetId) {
    long deadline = System.nanoTime() + TIMEOUT.toNanos();
    while (System.nanoTime() < deadline) {
      String sessionId = targetSessions.get(targetId);
      CompletableFuture<Void> configuration =
          sessionId == null ? null : sessionConfigurations.get(sessionId);
      if (configuration != null) {
        try {
          configuration.get(Math.max(1, deadline - System.nanoTime()), TimeUnit.NANOSECONDS);
          return sessionId;
        } catch (InterruptedException error) {
          Thread.currentThread().interrupt();
          throw new IllegalStateException("Interrupted activating CDP target", error);
        } catch (ExecutionException | TimeoutException error) {
          throw connectionFailure(new CompletionException(error));
        }
      }
      try {
        Thread.sleep(10);
      } catch (InterruptedException error) {
        Thread.currentThread().interrupt();
        throw new IllegalStateException("Interrupted activating CDP target", error);
      }
    }
    return "";
  }

  @Override
  public void close() {
    if (!closed.compareAndSet(false, true)) return;

    failPending(new IllegalStateException("CDP connection closed"));
    WebSocket current = socket;
    if (current != null) {
      current.sendClose(WebSocket.NORMAL_CLOSURE, "").exceptionally(error -> null);
    }
    targetSessions.clear();
    sessionConfigurations.clear();
  }

  @Override
  public void onOpen(WebSocket webSocket) {
    webSocket.request(1);
  }

  @Override
  public CompletableFuture<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
    String complete = null;
    synchronized (textLock) {
      textMessage.append(data);
      if (last) {
        complete = textMessage.toString();
        textMessage.setLength(0);
      }
    }

    if (complete != null) {
      try {
        onMessage(MAPPER.readTree(complete));
      } catch (JsonProcessingException e) {
        failConnection(new IllegalStateException("Could not parse CDP message", e));
      } catch (RuntimeException e) {
        failConnection(e);
      }
    }
    webSocket.request(1);
    return CompletableFuture.completedFuture(null);
  }

  @Override
  public CompletableFuture<?> onBinary(WebSocket webSocket, ByteBuffer data, boolean last) {
    failConnection(new IllegalStateException("Received an unexpected binary CDP message"));
    webSocket.request(1);
    return CompletableFuture.completedFuture(null);
  }

  @Override
  public CompletableFuture<?> onPing(WebSocket webSocket, ByteBuffer message) {
    webSocket.request(1);
    return webSocket.sendPong(message);
  }

  @Override
  public CompletableFuture<?> onPong(WebSocket webSocket, ByteBuffer message) {
    webSocket.request(1);
    return CompletableFuture.completedFuture(null);
  }

  @Override
  public CompletableFuture<?> onClose(WebSocket webSocket, int statusCode, String reason) {
    closed.set(true);
    failPending(new IllegalStateException("CDP connection closed"));
    return CompletableFuture.completedFuture(null);
  }

  @Override
  public void onError(WebSocket webSocket, Throwable error) {
    failConnection(new IllegalStateException("CDP connection failed", error));
  }

  private CompletableFuture<JsonNode> send(String method, Map<String, ?> params) {
    return send(method, params, "", true);
  }

  private CompletableFuture<JsonNode> send(
      String method, Map<String, ?> params, String sessionId, boolean wait) {
    if (closed.get()) {
      return CompletableFuture.failedFuture(new IllegalStateException("CDP connection closed"));
    }

    long id = nextId.getAndIncrement();
    Map<String, Object> message = new java.util.LinkedHashMap<>();
    message.put("id", id);
    message.put("method", method);
    message.put("params", params);
    if (!sessionId.isEmpty()) message.put("sessionId", sessionId);

    String json;
    try {
      json = MAPPER.writeValueAsString(message);
    } catch (JsonProcessingException e) {
      return CompletableFuture.failedFuture(e);
    }

    CompletableFuture<JsonNode> response = new CompletableFuture<>();
    if (wait) {
      pending.put(id, response);
      CompletableFuture.delayedExecutor(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS)
          .execute(
              () ->
                  response.completeExceptionally(
                      new TimeoutException("Timed out sending CDP command " + method)));
      response.whenComplete((result, error) -> pending.remove(id, response));
    }

    CompletableFuture<Void> written = enqueue(json);
    written.whenComplete(
        (ignored, error) -> {
          if (error != null) response.completeExceptionally(unwrap(error));
          else if (!wait) response.complete(MAPPER.createObjectNode());
        });
    return response;
  }

  private CompletableFuture<Void> enqueue(String message) {
    synchronized (sendLock) {
      sendTail =
          sendTail
              .handle((ignored, error) -> null)
              .thenCompose(
                  ignored -> {
                    WebSocket current = socket;
                    if (current == null || closed.get()) {
                      return CompletableFuture.failedFuture(
                          new IllegalStateException("CDP connection closed"));
                    }
                    return current.sendText(message, true).thenApply(webSocket -> null);
                  });
      return sendTail;
    }
  }

  private void onMessage(JsonNode message) {
    JsonNode idNode = message.get("id");
    if (idNode != null && idNode.canConvertToLong()) {
      CompletableFuture<JsonNode> command = pending.remove(idNode.longValue());
      if (command == null) return;

      JsonNode error = message.get("error");
      if (error != null && !error.isNull()) {
        String detail = error.path("message").asText("CDP command failed");
        command.completeExceptionally(new IllegalStateException(detail));
      } else {
        JsonNode result = message.get("result");
        command.complete(result == null ? MAPPER.createObjectNode() : result);
      }
      return;
    }

    String method = message.path("method").asText();
    JsonNode params = message.path("params");
    switch (method) {
      case "Target.attachedToTarget" -> attachedToTarget(params);
      case "Target.detachedFromTarget" -> detachedFromTarget(params.path("sessionId").asText());
      case "Target.targetCreated" -> {
        // Browser-level auto-attach handles page targets before their scripts run.
      }
      default -> {
        // Other CDP events are not relevant to page stability.
      }
    }
  }

  private void attachedToTarget(JsonNode params) {
    JsonNode target = params.path("targetInfo");
    String type = target.path("type").asText();
    String sessionId = params.path("sessionId").asText();
    if (sessionId.isEmpty()) return;
    CompletableFuture<Void> configuration = CompletableFuture.completedFuture(null);
    if ("page".equals(type) || "iframe".equals(type)) {
      String targetId = target.path("targetId").asText();
      if (!targetId.isEmpty()) targetSessions.put(targetId, sessionId);
      configuration = configureSession(sessionId);
    }

    configuration
        .handle(
            (result, error) -> {
              if (error != null) {
                LOG.debug("Could not configure CDP session {}", sessionId, unwrap(error));
              }
              return null;
            })
        .thenCompose(result -> send("Runtime.runIfWaitingForDebugger", Map.of(), sessionId, true))
        .exceptionally(
            error -> {
              LOG.debug("Could not resume CDP session {}", sessionId, unwrap(error));
              return null;
            });
  }

  private void detachedFromTarget(String sessionId) {
    if (sessionId.isEmpty()) return;

    sessionConfigurations.remove(sessionId);
    targetSessions.entrySet().removeIf(entry -> sessionId.equals(entry.getValue()));
  }

  private CompletableFuture<Void> configureSession(String sessionId) {
    if (sessionId == null || sessionId.isEmpty()) return CompletableFuture.completedFuture(null);
    return sessionConfigurations.computeIfAbsent(
        sessionId,
        ignored ->
            send("Target.setAutoAttach", AUTO_ATTACH_PARAMS, sessionId, true)
                .thenCompose(result -> send("Page.enable", Map.of(), sessionId, true))
                .thenCompose(
                    result ->
                        send(
                            "Page.addScriptToEvaluateOnNewDocument",
                            Map.of("source", waiterScript, "runImmediately", true),
                            sessionId,
                            true))
                .thenApply(result -> null));
  }

  private void failConnection(RuntimeException error) {
    closed.set(true);
    failPending(error);
    WebSocket current = socket;
    if (current != null) current.abort();
  }

  private void failPending(RuntimeException error) {
    pending.forEach(
        (id, command) -> {
          if (pending.remove(id, command)) command.completeExceptionally(error);
        });
  }

  private static URI websocketUrl(Capabilities capabilities) {
    Object cdp = capabilities.getCapability("se:cdp");
    if (cdp instanceof URI uri) return uri;
    if (cdp instanceof String value && !value.isBlank()) return URI.create(value);

    Object address = debuggerAddress(capabilities.getCapability("goog:chromeOptions"));
    if (address == null) {
      address = debuggerAddress(capabilities.getCapability("ms:edgeOptions"));
    }
    if (address == null || address.toString().isBlank()) {
      throw new IllegalStateException("Chromium did not expose a CDP debugger address");
    }

    URI versionUrl = URI.create("http://" + address + "/json/version");
    HttpClient client = HttpClient.newBuilder().connectTimeout(TIMEOUT).build();
    HttpRequest request = HttpRequest.newBuilder(versionUrl).timeout(TIMEOUT).GET().build();
    try {
      HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
      if (response.statusCode() < 200 || response.statusCode() >= 300) {
        throw new IllegalStateException("CDP discovery failed: " + response.statusCode());
      }
      String url = MAPPER.readTree(response.body()).path("webSocketDebuggerUrl").asText();
      if (url.isEmpty()) {
        throw new IllegalStateException("CDP discovery returned no webSocketDebuggerUrl");
      }
      return URI.create(url);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new IllegalStateException("CDP discovery was interrupted", e);
    } catch (IOException e) {
      throw new IllegalStateException("CDP discovery failed", e);
    }
  }

  private static Object debuggerAddress(Object options) {
    return options instanceof Map<?, ?> map ? map.get("debuggerAddress") : null;
  }

  private static RuntimeException connectionFailure(RuntimeException error) {
    Throwable cause = unwrap(error);
    return cause instanceof RuntimeException runtime
        ? runtime
        : new IllegalStateException("Could not connect to Chromium CDP", cause);
  }

  private static Throwable unwrap(Throwable error) {
    Throwable current = error;
    while ((current instanceof CompletionException
            || current instanceof java.util.concurrent.ExecutionException)
        && current.getCause() != null) {
      current = current.getCause();
    }
    return current;
  }
}
